File: src/my_py_pkg/my_py_pkg/add_2_ints_client.py
#!/usr/bin/env python3
#

import rclpy
from rclpy.node import Node
from rclpy.task import Future

# ros2 interface show example_interfaces/srv/AddTwoInts
# AddTwoInts.a <- name of the first op. 
# AddTwoInts.b <- name of the second op.
# AddTwoInts.sum <- response sum = a + b.
from example_interfaces.srv import AddTwoInts
from functools import partial 

SERVICE_NAME = "add_2_ints"

class Add2IntsClientNode( Node ):
    def __init__( self ):
        super().__init__( "add_2_ints_client" )
        self.client_ = self.create_client( AddTwoInts, SERVICE_NAME )
        self.get_logger().info( "ADD2IntsClient has been started." )

    def call_add_2_ints( self, a, b ):
        while not self.client_.wait_for_service( 1.0 ):
            self.get_logger().warn( "Waiting for AddTwoInts server." )
        
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = self.client_.call_async( request )
        future.add_done_callback( 
            partial( self.callback_call_add_2_ints, request=request ) )

    def callback_call_add_2_ints( self, future: Future,
                                  request: AddTwoInts.Request ):
        response: AddTwoInts.Response = future.result()
        self.get_logger().info( str( request.a ) + " + " + 
                                str( request.b ) +
                                " = " + str( response.sum ) )


def main( args = None ):
    rclpy.init( args = args )
    node = Add2IntsClientNode()
    node.call_add_2_ints( 2, 7 )
    node.call_add_2_ints( -2, 2 )
    node.call_add_2_ints( -4, 10 )

    rclpy.spin( node )
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_2_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
/*
# ros2 interface show example_interfaces/srv/AddTwoInts
# AddTwoInts.a <- name of the first op. 
# AddTwoInts.b <- name of the second op.
# AddTwoInts.sum <- response sum = a + b.
*/
#include "example_interfaces/srv/add_two_ints.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;

#define SERVICE_NAME "add_2_ints"

class Add2IntsClientNode : public rclcpp::Node
{
    public:
        Add2IntsClientNode() : Node( "add_2_ints_client" )
        {
            this->client_ = this->create_client<example_interfaces::srv::AddTwoInts>( SERVICE_NAME );
            RCLCPP_INFO( this->get_logger(), "Add2IntsClient has been started." );
        };

        void callAdd2Ints( int64_t a, int64_t b )
        {
            while( !this->client_->wait_for_service( 1s ) ) {
                RCLCPP_WARN( this->get_logger(), "Waiting for AddTwoInts server." );
            }

            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = a;
            request->b = b;
            this->client_->async_send_request( request,
                std::bind( &Add2IntsClientNode::callbackCallAdd2Ints, this, _1 ) );
        };

    private:
        rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;

        void callbackCallAdd2Ints( rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future )
        {
            auto response = future.get();
            RCLCPP_INFO( this->get_logger(), "%ld", response->sum );
        };
};

int main( int argc, char **argv )
{
    rclcpp::init( argc, argv );
    auto node = std::make_shared<Add2IntsClientNode>();
    node->callAdd2Ints( 1, 3 );
    node->callAdd2Ints( -3, 3 );
    node->callAdd2Ints( -1, -3 );
    rclcpp::spin( node );
    rclcpp::shutdown();
    return 0;
};
